v2/workloadapi/client.go

```diff
@@ -2,75 +2,411 @@ package workloadapi
+		return nil, err
+	}
+	bundles := []*x509bundle.Bundle{}
+	for _, svid := range resp.Svids {
+		td, err := spiffeid.TrustDomainFromString(svid.SpiffeId)
+		if err != nil {
+			return nil, err
+		}
+		certs, err := x509.ParseCertificates(svid.Bundle)
+		if err != nil {
+			return nil, err
+		}
+		b := x509bundle.FromX509Roots(td, certs)
+		bundles = append(bundles, b)
+	}
+	return x509bundle.NewSet(bundles...), nil
 }
 
 // FetchX509Context fetches the X.509 context, which contains both X509-SVIDs
 // and X.509 bundles.
 func (c *Client) FetchX509Context(ctx context.Context) (*X509Context, error) {
-	panic("not implemented")
+	ctx, cancel := context.WithCancel(withHeader(ctx))
+	defer cancel()
+	stream, err := c.wlClient.FetchX509SVID(ctx, &workload.X509SVIDRequest{})
+	if err != nil {
+		return nil, err
+	}
+	resp, err := stream.Recv()
+	if err != nil {
+		return nil, err
+	}
+	svids := []*x509svid.SVID{}
+	bundles := []*x509bundle.Bundle{}
+	for _, svid := range resp.Svids {
+		s, err := x509svid.ParseRaw(svid.X509Svid, svid.X509SvidKey)
+		if err != nil {
+			return nil, err
+		}
+		svids = append(svids, s)
+
+		certs, err := x509.ParseCertificates(svid.Bundle)
+		if err != nil {
+			return nil, err
+		}
+		b := x509bundle.FromX509Roots(s.ID.TrustDomain(), certs)
+		bundles = append(bundles, b)
+	}
+	return &X509Context{
+		SVIDs:   svids,
+		Bundles: x509bundle.NewSet(bundles...),
+	}, nil
 }
 
 // WatchX509Context watches for updates to the X.509 context. The watcher
 // receives the updated X.509 context.
 func (c *Client) WatchX509Context(ctx context.Context, watcher X509ContextWatcher) error {
-	panic("not implemented")
+	for {
+		err := c.watchX509Context(ctx, watcher)
+		watcher.OnX509ContextWatchError(err)
+		err = c.handleCallError(ctx, err)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (c *Client) handleCallError(ctx context.Context, err error) error {
+	code := status.Code(err)
+	if code == codes.Canceled || code == codes.InvalidArgument {
+		c.opts.log.Errorf("cancelling : %v", err)
+		return err
+	}
+	c.opts.log.Errorf("there was an error with the gRPC call: %v", err)
+	c.opts.log.Infof("backing off before retrying")
+
+	select {
+	case <-time.After(c.backoff.Duration()):
+		c.opts.log.Infof("retrying")
+		return nil
+
+	case <-ctx.Done():
+		c.opts.log.Infof("context was done: %v", ctx.Err())
+		return ctx.Err()
+	}
+}
+
+func (c *Client) watchX509Context(ctx context.Context, watcher X509ContextWatcher) error {
+	ctx, cancel := context.WithCancel(withHeader(ctx))
+	defer cancel()
+
+	stream, err := c.wlClient.FetchX509SVID(ctx, &workload.X509SVIDRequest{})
+	if err != nil {
+		return err
+	}
+
+	for {
+		resp, err := stream.Recv()
+		if err != nil {
+			return err
+		}
+
+		svids := []*x509svid.SVID{}
+		bundles := []*x509bundle.Bundle{}
+		for _, svid := range resp.Svids {
+			s, err := x509svid.ParseRaw(svid.X509Svid, svid.X509SvidKey)
+			if err != nil {
+				watcher.OnX509ContextWatchError(err)
+				continue
```

Comment:
I think we should throw out the whole response if we encounter something wrong with any part of it.